src/H5HL.c

```diff
@@ -508,27 +508,33 @@ END_FUNC(STATIC) /* end H5HL__dirty() */
  *
  * Purpose:     Inserts a new item into the heap.
  *
- * Return:      Success:    Offset of new item within heap.
- *              Failure:    UFAIL
+ * Return:      Success:    SUCCEED
+ *                          Offset set to location of new item within heap
+ *
+ *              Failure:    FAIL
+ *                          Offset set to SIZE_MAX
  *
  * Programmer:  Robb Matzke
  *              Jul 17 1997
  *
  *-------------------------------------------------------------------------
  */
-BEGIN_FUNC(PRIV, ERR, size_t, UFAIL, UFAIL,
-           H5HL_insert(H5F_t *f, H5HL_t *heap, size_t buf_size, const void *buf))
+BEGIN_FUNC(PRIV, ERR, herr_t, SUCCEED, FAIL,
+           H5HL_insert(H5F_t *f, H5HL_t *heap, size_t buf_size, const void *buf, size_t *offset_out))
 
     H5HL_free_t *fl = NULL, *last_fl = NULL;
-    size_t       offset = 0;
     size_t       need_size;
+    size_t       offset = 0;
     hbool_t      found;
 
-    /* check arguments */
+    /* Check arguments */
     HDassert(f);
     HDassert(heap);
     HDassert(buf_size > 0);
     HDassert(buf);
+    HDassert(offset_out);
+
+    *offset_out = SIZE_MAX;
```

Comment:
I would avoid touching *offset_out until there's no possibility for errors, so you don't modify the user's state when if an error does occur.  As much as possible, we should try to have "atomic" operations that only modify state on success.

Reply:
I went ahead and changed this, but we should probably think more carefully about what policy we want. Immediately setting output values to "likely to fail" values is probably more robust when people pass in pointers to uninitialized state and then don't bother to check error codes.